roachprod: add Grow command

herkolategan opened the pull request:
Previously, after creating a managed (MIG) cluster in GCE it was not possible to add nodes. This change introduces a new `Grow` method on the gcloud Provider, and as part of the Provider interface.

New nodes are distributed among the zones, the groups are currently in, to balance out the nodes.

Epic: CRDB-33832
See: #117802

Release Note: None

herkolategan committed:
roachprod: add Grow capability to gcloud

Previously, after creating a managed (MIG) cluster in GCE it was not possible to
add nodes. This change introduces a new `Grow` method on the gcloud Provider,
and as part of the Provider interface. When growing a cluster, new nodes are
distributed among the groups balance out the nodes.

Epic: CRDB-33832
See: #117802

Release Note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118912)
<!-- Reviewable:end -->

herkolategan committed:
roachprod: add Grow command to roachprod cli

Expose Grow Cluster on the `roachprod` CLI tool.

Epic: CRDB-33832
Resolves: #117802

Release Note: None

DarrylWong commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1445,7 +1445,88 @@ func (p *Provider) Create(
+		createArgs := []string{"compute", "instance-groups", "managed", "create-instance", "--zone", group.Zone, groupName}
+		for i := 0; i < addCount; i++ {
+			name := names[i]
+			names = names[1:]
```
It makes sense that we remove the first element of `names`, that vm has been added. But in that case, shouldn't the name we are adding always be `name[0]`?

i.e. we have the instance groups with size [1, 1, 1]

We call Grow(4).

`names = [vm-1, vm-2, vm-3, vm-4]`

Iteration 1: `name=vm-1`, `names=[vm-2, vm-3, vm-4]`

Iteration 2:  `name=vm-3`, `names=[vm-3, vm-4]`, which is wrong since we removed `vm-2` but didn't actually add it yet.

DarrylWong commented on pkg/roachprod/vm/gce/gcloud.go:
```diff
@@ -1445,7 +1445,88 @@ func (p *Provider) Create(
+}
+
+// computeGrowDistribution computes the distribution of new nodes across the
+// existing instance groups. The distribution is computed naively, for
```
Just curious, what would be a less naive way of distribution? Avoiding all the nodes in a cluster getting placed in the same instance group?

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118912)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1452 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118912#gh-1496649396:-NrA4g1c6eC5F9Se36_q:b-a0ylwm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d5d0c9b2cc628866a7a6736637f34abc69e91e/pkg/roachprod/vm/gce/gcloud.go#L1452)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

Just curious, what would be a less naive way of distribution? Avoiding all the nodes in a cluster getting placed in the same instance group?
</blockquote></details>

I think I mostly meant naive in the sense that it's not optimised for the likelihood of there being more nodes to distribute than there are instance groups. For instance if we distributed N nodes, we'd need to do N operations, even if there were only Y instance groups to distribute to (ex. 1000 nodes across 3 instance groups). There's a mathematical way of doing it, I think, which might be computationally less intensive, but results in more complex code / edge cases.

But given that this is not a hot path operation, optimisation is by far not required :)
"naively" may be a strong word here.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1498 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118912#gh-1496636173:-NrA60Z86zXV0C4XFqW1:byni21f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d5d0c9b2cc628866a7a6736637f34abc69e91e/pkg/roachprod/vm/gce/gcloud.go#L1498)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

It makes sense that we remove the first element of `names`, that vm has been added. But in that case, shouldn't the name we are adding always be `name[0]`?

i.e. we have the instance groups with size [1, 1, 1]

We call Grow(4).

`names = [vm-1, vm-2, vm-3, vm-4]`

Iteration 1: `name=vm-1`, `names=[vm-2, vm-3, vm-4]`

Iteration 2:  `name=vm-3`, `names=[vm-3, vm-4]`, which is wrong since we removed `vm-2` but didn't actually add it yet.
</blockquote></details>

Good catch! that one slipped by me; Refactoring Error gone rogue... Some form of a dry-run / intercepted interface would do nicely to test this code, and would have caught this 🤔.

```
panic: runtime error: index out of range [3] with length 2
```


<!-- Sent from Reviewable.io -->

DarrylWong reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118912)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, and @srosenberg)

___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1452 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118912#gh-1496649396:-NrB9HSY6O8g0DJWDpHP:bb6ehrp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d5d0c9b2cc628866a7a6736637f34abc69e91e/pkg/roachprod/vm/gce/gcloud.go#L1452)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

I think I mostly meant naive in the sense that it's not optimised for the likelihood of there being more nodes to distribute than there are instance groups. For instance if we distributed N nodes, we'd need to do N operations, even if there were only Y instance groups to distribute to (ex. 1000 nodes across 3 instance groups). There's a mathematical way of doing it, I think, which might be computationally less intensive, but results in more complex code / edge cases.

But given that this is not a hot path operation, optimisation is by far not required :)
"naively" may be a strong word here.
</blockquote></details>

Got it! I think the wording was fine, my brain was just fixated on figuring out why adding to the smallest group was naive and completely missed the fact that it's because we're doing so one by one. Anyway, I agree it's fine being unoptimized in this case.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1498 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118912#gh-1496636173:-NrBBiCyD1FKBsWl2YKa:b4e5tat) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d5d0c9b2cc628866a7a6736637f34abc69e91e/pkg/roachprod/vm/gce/gcloud.go#L1498)):*
<details><summary><i>Previously, herkolategan (Herko Lategan) wrote…</i></summary><blockquote>

Good catch! that one slipped by me; Refactoring Error gone rogue... Some form of a dry-run / intercepted interface would do nicely to test this code, and would have caught this 🤔.

```
panic: runtime error: index out of range [3] with length 2
```
</blockquote></details>

Yeah I think at least a dry run to confirm everything works as expected would be good, but besides that it LGTM.


<!-- Sent from Reviewable.io -->

renatolabs reviewed (approved):
Nice! My comments are all optional/nits.

General notes:

* Usability-wise, it would be nice to have _some_ kind of feedback sent to the user with `roachprod grow`. It's silent for a few seconds, and only outputs something when the disk propagation logic kicks in.
* Do we envision being able to provision a new zone? That would be a nice feature, but can definitely be follow up / future work.

> Reviewed 5 of 8 files at r1, 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-:-NrBMvuT6WOE__FwLem_:b-rij3d8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @herkolategan, and @srosenberg)

___
*[`-- commits` line 9 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrB9ng-1SC4CWrY1nb9):*
Nit: this sentence is a little weird?
___
*[`pkg/roachprod/roachprod.go` line 1518 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrBMpQgBFeRNfZlyol-:-NrBMpQgBFeRNfZlyol0:b-k3egwj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f762346559424de7a9086254f0bdc0d1835059ba/pkg/roachprod/roachprod.go#L1518)):*
> ```Go
> 	for _, cVM := range c.VMs {
> 		if cVM.Provider != gce.ProviderName {
> 			return fmt.Errorf("growing a cluster is currently only supported on gce")
> ```

Should we remove this check, since `cloud.GrowCluster` is already doing it?
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1452 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118912#gh-1496649396:-NrBJMQM7piS7dgWouHw:b-hmq284) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d5d0c9b2cc628866a7a6736637f34abc69e91e/pkg/roachprod/vm/gce/gcloud.go#L1452)):*
<details><summary><i>Previously, DarrylWong wrote…</i></summary><blockquote>

Got it! I think the wording was fine, my brain was just fixated on figuring out why adding to the smallest group was naive and completely missed the fact that it's because we're doing so one by one. Anyway, I agree it's fine being unoptimized in this case.
</blockquote></details>

Even though it's "naive", it took me a few minutes of thinking about indexes in slices to convince myself that this function works, heh.

```go
addCount := make([]int, len(groups))

// every zone gets the same amount of nodes
newNodesPerZone := newNodeCount / len(groups)
for j := range addCount {
    addCount[j] = newNodesPerZone
}

// remaining nodes are distributed evenly across smaller zones
remainingNodes := newNodeCount % len(groups)
for j := 0; j < remainingNodes; j++ {
    addCount[j]++
}
```

This is totally untested, but would something like it work? It does assume that the zones are balanced (or off by at most one) but AFAIK that's a roachprod guarantee anyway.

This is mostly for readability, not for performance or optimization. At the end of the day, the real "performance" would come by minimizing `gcloud` calls but, as far as I can tell, there's no way to create multiple instances with a single `gcloud` call, so anything we do here "doesn't matter".

This is totally optional and there are likely edge cases I haven't considered, as you suggested.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1470 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrBJ9be3bRPLkdHtbk7:-NrBJ9be3bRPLkdHtbk8:bsn278n) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c15fc1668dfa578e8c515f53a6d84ba4465b5e4/pkg/roachprod/vm/gce/gcloud.go#L1470)):*
> ```Go
> 			if curIndex >= len(groups) {
> 				curIndex = 0
> 			}
> ```

Isn't this just `curIndex = nextIndex`?
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1488 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrBLvCw-EIVw6nt-sKl:-NrBLvCw-EIVw6nt-sKm:b8a82zj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c15fc1668dfa578e8c515f53a6d84ba4465b5e4/pkg/roachprod/vm/gce/gcloud.go#L1488)):*
> ```Go
> 
> 	newNodeCount := len(names)
> 	addCounts := computeGrowDistribution(groups, newNodeCount)
> ```

I'd move the sorting outside of this function. It's an unexpected side effect, I it would be clearer to have that in this function and make the sorting a precondition for `computeGrowDistribution`.
___
*[`pkg/roachprod/vm/gce/gcloud_test.go` line 259 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrBMERA2sgKZNZo086Q:-NrBMERA2sgKZNZo086R:bb66cmv) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c15fc1668dfa578e8c515f53a6d84ba4465b5e4/pkg/roachprod/vm/gce/gcloud_test.go#L259)):*
> ```Go
> 		return true
> 	}
> 	if err := quick.Check(testDistribution, &c); err != nil {
> ```

Nice! :clap: 


<!-- Sent from Reviewable.io -->

herkolategan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118912)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DarrylWong, @renatolabs, and @srosenberg)

___
*[`-- commits` line 9 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrB9ng-1SC4CWrY1nb9):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Nit: this sentence is a little weird?
</blockquote></details>

true, I'll just mention groups to keep it simple.
___
*[`pkg/roachprod/roachprod.go` line 1518 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrBMpQgBFeRNfZlyol-:-NrF77uy81nGnIu29IO4:b-93i47d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f762346559424de7a9086254f0bdc0d1835059ba/pkg/roachprod/roachprod.go#L1518)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Should we remove this check, since `cloud.GrowCluster` is already doing it?
</blockquote></details>

yep, it's redundant.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1452 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/118912#gh-1496649396:-NrF1Cw47nsOY04-NyEs:bcmio88) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5d5d0c9b2cc628866a7a6736637f34abc69e91e/pkg/roachprod/vm/gce/gcloud.go#L1452)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Even though it's "naive", it took me a few minutes of thinking about indexes in slices to convince myself that this function works, heh.

```go
addCount := make([]int, len(groups))

// every zone gets the same amount of nodes
newNodesPerZone := newNodeCount / len(groups)
for j := range addCount {
    addCount[j] = newNodesPerZone
}

// remaining nodes are distributed evenly across smaller zones
remainingNodes := newNodeCount % len(groups)
for j := 0; j < remainingNodes; j++ {
    addCount[j]++
}
```

This is totally untested, but would something like it work? It does assume that the zones are balanced (or off by at most one) but AFAIK that's a roachprod guarantee anyway.

This is mostly for readability, not for performance or optimization. At the end of the day, the real "performance" would come by minimizing `gcloud` calls but, as far as I can tell, there's no way to create multiple instances with a single `gcloud` call, so anything we do here "doesn't matter".

This is totally optional and there are likely edge cases I haven't considered, as you suggested.
</blockquote></details>

Yes, I initially had something similar to this. And it likely will work as long as the groups are already balanced with a max variance of 1 (odd node out). Then I thought about the scenario we mentioned above where we may want to add new groups / zones / regions. But even then we'd probably want to give options in regards to the distribution. For now I'll keep the balancing algorithm for future use. I'm likely to revisit this at some point, and then we could add a flag for redistribution algo to use.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1470 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrBJ9be3bRPLkdHtbk7:-NrF2EkEAQL0A2_r30XC:bq0wma6) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c15fc1668dfa578e8c515f53a6d84ba4465b5e4/pkg/roachprod/vm/gce/gcloud.go#L1470)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Isn't this just `curIndex = nextIndex`?
</blockquote></details>

I believe it is, I'll simplify.
___
*[`pkg/roachprod/vm/gce/gcloud.go` line 1488 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118912#-NrBLvCw-EIVw6nt-sKl:-NrF2IeA4g4gJotm9Dyp:b-umdj1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c15fc1668dfa578e8c515f53a6d84ba4465b5e4/pkg/roachprod/vm/gce/gcloud.go#L1488)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I'd move the sorting outside of this function. It's an unexpected side effect, I it would be clearer to have that in this function and make the sorting a precondition for `computeGrowDistribution`.
</blockquote></details>

Yeah, I had the same feeling, I'll add a comment to the function then that the passed groups must be ordered (it does however make me want to add a check in the function to ensure it is so).


<!-- Sent from Reviewable.io -->

herkolategan commented:
TFTRs!

bors r=DarrylWong,renatolabs
